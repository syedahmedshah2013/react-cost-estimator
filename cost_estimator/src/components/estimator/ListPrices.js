import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';

import { FetchPrices } from '../../actions';

class ListPrices extends Component {
    state = { pricesStatus: 0, prices: [], favorite: false }

    componentDidMount() {
        this.fetchItemPrices();
    }

    fetchItemPrices = async () => {
        await this.props.FetchPrices();
        console.log(this.props.prices);
        
        if(this.props.prices && this.props.prices.cost.items.length > 0) {
            this.setState({ pricesStatus: 200, prices: this.props.prices.cost.items })
        } else {
            this.setState({ pricesStatus: 500 });
        }
    }

    renderListItems = () => {
        if(this.state.pricesStatus === 200) {
            if(this.state.prices.length > 0) {
                return (
                    <tbody>
                        {this.state.prices.map(({ name, net, tax }, index) => {
                            let gross = parseFloat(net) + parseFloat(tax);
                            return (
                                <Fragment>
                                    <tr>
                                        <th scope="row">{name}</th>
                                        <td>{net} €</td>
                                        <td>{tax} €</td>
                                        <td>{gross} €</td>
                                        <td className="d-flex justify-content-between">
                                            <div className=""><i className="bi bi-x-lg"></i></div>
                                            <div><i className="bi bi-star"></i></div>
                                        </td>
                                    </tr>
                                </Fragment>
                            );
                        })}
                    </tbody>
                )
            } else {
                return (
                    <div>Oops! No items found.</div>
                );
            }
        } else {
            return (
                <div>Oops! Something went wrong while listing records.</div>
            )
        }
    }

    render() {
        return(
            <Fragment>
                <table className="table">
                    <thead>
                        <tr>
                            <th scope="col">Name</th>
                            <th scope="col">Net</th>
                            <th scope="col">Tax</th>
                            <th scope="col">Gross</th>
                            <th scope="col"></th>
                        </tr>
                    </thead>
                    
                    {this.renderListItems()}
                </table>
            </Fragment>
        );
    }
}

const mapStateToProps = ({ prices }) => {
    return { prices }
}

export default connect(mapStateToProps, { FetchPrices })(ListPrices);