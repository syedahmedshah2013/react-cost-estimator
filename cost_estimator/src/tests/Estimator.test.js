import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import toJson from "enzyme-to-json";

import ListPrices from '../components/estimator/ListPrices';

const mockStore = configureStore([]);

const initState = {
    prices: {
      status: 200,
      type: 'success',
      msg: 'Successfully fetched the prices.',
      cost: {
        items: [
          {
            id: 1,
            name: 'Architect',
            net: '1000.0',
            tax: '160.0'
          },
          {
            id: 2,
            name: 'Bricks',
            net: '1500.32',
            tax: '240.0512'
          },
          {
            id: 3,
            name: 'Property',
            net: '2000.5',
            tax: '320.08'
          },
          {
            id: null,
            name: 'HR',
            net: '2500',
            tax: 400,
            gross: 2900
          },
          {
            id: null,
            name: 'Accountant',
            net: '3500',
            tax: 560,
            gross: 4060
          }
        ]
      }
    }
}

describe('<ListPrices/>', () => {
    let store;
    let component;

    beforeEach(() => {
        store = mockStore({
            initState: initState
        });
        component = renderer.create(
            <Provider store={store}>
                <ListPrices />
            </Provider>
        )
    });

    it('ListPrices component renders with given state from Redux Store', () => {
        expect(component.toJSON()).toMatchSnapshot();
    });
})