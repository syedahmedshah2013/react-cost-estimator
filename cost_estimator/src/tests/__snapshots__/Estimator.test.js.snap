// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListPrices/> ListPrices component renders with given state from Redux Store 1`] = `
Array [
  <div
    className="row"
  >
    <div
      className="col-12"
    >
      <button
        className="btn btn-sm btn-success"
        onClick={[Function]}
      >
        Show Favorites
      </button>
    </div>
  </div>,
  <table
    className="table"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          Name
        </th>
        <th
          scope="col"
        >
          Net
        </th>
        <th
          scope="col"
        >
          Tax
        </th>
        <th
          scope="col"
        >
          Gross
        </th>
        <th
          scope="col"
        />
      </tr>
    </thead>
    <div>
      Oops! Something went wrong while listing records.
    </div>
  </table>,
]
`;

exports[`<ListPrices/> ListPrices component renders with given store 1`] = `
Array [
  <div
    className="row"
  >
    <div
      className="col-12"
    >
      <button
        className="btn btn-sm btn-success"
        onClick={[Function]}
      >
        Show Favorites
      </button>
    </div>
  </div>,
  <table
    className="table"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          Name
        </th>
        <th
          scope="col"
        >
          Net
        </th>
        <th
          scope="col"
        >
          Tax
        </th>
        <th
          scope="col"
        >
          Gross
        </th>
        <th
          scope="col"
        />
      </tr>
    </thead>
    <div>
      Oops! Something went wrong while listing records.
    </div>
  </table>,
]
`;
